adc/consumer.py

```diff
@@ -0,0 +1,180 @@
+        """
+        self.logger.debug(f"subscribing to topic {topic}")
+
+        try:
+            topic_meta = self.describe_topic(topic, timeout)
+        except KeyError:
+            raise ValueError(f"topic {topic} does not exist on the broker, so can't subscribe")
+
+        assignment = []
+        for partition_id in topic_meta.partitions.keys():
+            self.logger.debug(f"adding subscription to topic partition={partition_id}")
+            tp = confluent_kafka.TopicPartition(
+                topic=topic,
+                partition=partition_id,
+                offset=confluent_kafka.OFFSET_BEGINNING,
+            )
+            assignment.append(tp)
+
+        self.logger.debug("registering topic assignment")
+        self._consumer.assign(assignment)
+
+    def describe_topic(
+            self,
+            topic: str,
+            timeout: timedelta = timedelta(seconds=5.0)) -> confluent_kafka.admin.TopicMetadata:
+        """Fetch confluent_kafka.admin.TopicMetadata describing a topic.
+        """
+        self.logger.debug(f"fetching cluster metadata to describe topic name={topic}")
+        cluster_meta = self._consumer.list_topics(timeout=timeout.total_seconds())
+        self.logger.debug(f"cluster metadata: {cluster_meta.topics}")
+        return cluster_meta.topics[topic]
+
+    def mark_done(self, msg: confluent_kafka.Message):
+        """
+        Mark a message as fully-processed. In the background, the client will
+        continuously synchronize this information with Kafka so that the stream can be
+        resumed from this point in the future.
+        """
+        self._consumer.store_offsets(msg)
+
+    def message_stream(self,
+                       autocommit: bool = True,
+                       batch_size: int = 100,
+                       batch_timeout: timedelta = timedelta(seconds=1.0),
+                       ) -> Iterator[confluent_kafka.Message]:
+
+        """Returns a stream which iterates over the messages in the topics
+        to which the client is subscribed.
+
+        If autocommit is true, then messages are automatically marked as handled
+        when the next message is yielded. This removes the need to call
+        'mark_done' on each message. Callers using asynchronous message
+        processing or with complex processing needs should disable this.
+
+        batch_size controls the number of messages to request from Kafka per
+        batch. Higher values may be more efficient, but may add latency.
+
+        batch_timeout controls how long the client should wait for Kafka to
+        provide a full batch of batch_size messages. Higher values may be more
+        efficient, but may add latency.
+
+        The stream stops when the client has hit the last message in all
+        partitions. This set of partitions is calculated just once when
+        iterate() is first called; calling subscribe() after iterate() may
+        cause inconsistent behavior.
```

Comment:
Yep, should be `message_stream`.

Reply:
I wonder if it would be better to call that `stream()` instead (short methods for commonly-used behavior) or skip that entirely and just provide an `__iter__` method so you can iterate messages by calling the Consumer, similarly to how `KafkaConsumer` operates in kafka-python:

```
c = Consumer(cfg)
c.subscribe("topic1")
c.subscribe("topic2")
for msg in c:
    print(msg.value())
    print(msg.topic())
```

Reply:
You know, as soon as one looks at `kafka-python`, one sort of wishes we were just using it directly...

Reply:
I renamed it `stream()`, but didn't go as far as `__iter__`. It just seems odd to me to say `for msg in consumer` - the messages aren't in a _consumer_, they're in a stream that the consumer is consuming.

Reply:
Agreed, that sounds perfectly reasonable.